service/resource/deploymentv3/desired_test.go

```diff
@@ -92,17 +100,17 @@ func Test_Resource_Deployment_GetDesiredState(t *testing.T) {
 							StorageType: "hostPath",
 						},
 						Masters: []v1alpha1.KVMConfigSpecKVMNode{
-							{CPUs: 1, Memory: "1Gi"},
+							{CPUs: 1, Memory: "1G"},
 						},
 						NodeController: v1alpha1.KVMConfigSpecKVMNodeController{
 							Docker: v1alpha1.KVMConfigSpecKVMNodeControllerDocker{
 								Image: "123",
 							},
 						},
 						Workers: []v1alpha1.KVMConfigSpecKVMNode{
-							{CPUs: 4, Memory: "8.2G"},
-							{CPUs: 4, Memory: "8.2G"},
-							{CPUs: 4, Memory: "8.2G"},
+							{CPUs: 4, Memory: "8G"},
+							{CPUs: 4, Memory: "8G"},
+							{CPUs: 4, Memory: "8G"},
```

Comment:
Does that mean decimals do not work? Did they ever work?

Reply:
AFAIK they should work, I changed this mostly to keep tests readable. Memory values are represented as `Quantity` https://github.com/kubernetes/apimachinery/blob/master/pkg/api/resource/quantity.go#L35, we use the `Add` method https://github.com/kubernetes/apimachinery/blob/master/pkg/api/resource/quantity.go#L546 to increment the given value in 1G. `reflect.DeepEqual` doesn't work well with `Quantity.Add` because the underlying values can be the same but the representation can be different and the struct comparison can fail. The alternative approach would be to unpack the structs in the test, extract the actual Quantity values and compare, it was getting really ugly when I switched to using simpler test values.

Reply:
I wanted to test decimal in the first place. I would argue it improves readability.

Reply:
I'd be fine with comparing string representation of the values with a comment why you had to do that.

Reply:
After calling key.MemoryQuantity the input string is changed into a Quantity. This type gives a useful method for adding, we can increase our input Quantity by 1G no matter how it was represented, the final value will be ok. However, after adding, the Quantity that represents that value is different than what we get from `resource.MustParse` in the tests for decimal or Gi values incremented by 1G. And `reflect.DeepEqual` fails:
```
--- FAIL: Test_Resource_Deployment_GetDesiredState (0.00s)
	desired_test.go:317: case 1 expected []v1.ResourceRequirements{v1.ResourceRequirements{Limits:v1.ResourceList{"cpu":resource.Quantity{i:resource.int64Amount{value:4, scale:0}, d:resource.infDecAmount{Dec:(*inf.Dec)(nil)}, s:"4", Format:"DecimalSI"}, "memory":resource.Quantity{i:resource.int64Amount{value:92, scale:8}, d:resource.infDecAmount{Dec:(*inf.Dec)(nil)}, s:"", Format:"DecimalSI"}}, Requests:v1.ResourceList{"cpu":resource.Quantity{i:resource.int64Amount{value:4, scale:0}, d:resource.infDecAmount{Dec:(*inf.Dec)(nil)}, s:"4", Format:"DecimalSI"}, "memory":resource.Quantity{i:resource.int64Amount{value:92, scale:8}, d:resource.infDecAmount{Dec:(*inf.Dec)(nil)}, s:"", Format:"DecimalSI"}}}} got []v1.ResourceRequirements{v1.ResourceRequirements{Limits:v1.ResourceList{"cpu":resource.Quantity{i:resource.int64Amount{value:4, scale:0}, d:resource.infDecAmount{Dec:(*inf.Dec)(nil)}, s:"4", Format:"DecimalSI"}, "memory":resource.Quantity{i:resource.int64Amount{value:92, scale:8}, d:resource.infDecAmount{Dec:(*inf.Dec)(nil)}, s:"9200M", Format:"DecimalSI"}}, Requests:v1.ResourceList{"cpu":resource.Quantity{i:resource.int64Amount{value:4, scale:0}, d:resource.infDecAmount{Dec:(*inf.Dec)(nil)}, s:"4", Format:"DecimalSI"}, "memory":resource.Quantity{i:resource.int64Amount{value:92, scale:8}, d:resource.infDecAmount{Dec:(*inf.Dec)(nil)}, s:"9200M", Format:"DecimalSI"}}}}
```
Comparing string representations would be the same as comparing values, we are dealing with Quantities and we need to extract the string representations after the increase using Quantity methods, in this case https://github.com/kubernetes/apimachinery/blob/master/pkg/api/resource/quantity.go#L609, again not as readable as `reflect.DeepEqual` IMO because of how we need to extract both the expected and actual values, etc.

Reply:
You convinced me we should skip decimal tests ;)